src/gd_avif.c

```diff
@@ -0,0 +1,667 @@
+
+    speed - At slower speeds, encoding may be quite slow. Use judiciously.
+
+  Returns:
+
+    Nothing.
+*/
+
+BGD_DECLARE(void) gdImageAvifEx (gdImagePtr im, FILE * outFile, int quality, int speed)
+{
+	gdIOCtx *out = gdNewFileCtx(outFile);
+
+	if (out == NULL) {
+		return;
+	}
+
+	gdImageAvifCtx(im, out, quality, speed);
+	out->gd_free(out);
+}
+
+BGD_DECLARE(void) gdImageAvif (gdImagePtr im, FILE * outFile)
+{
+	return gdImageAvifEx(im, outFile, DEFAULT_QUALITY, AVIF_SPEED_DEFAULT);
+}
+
+BGD_DECLARE(void *) gdImageAvifPtrEx (gdImagePtr im, int *size, int quality, int speed)
+{
+	void * rv;
+	gdIOCtx * out = gdNewDynamicCtx(NEW_DYNAMIC_CTX_SIZE, NULL);
+
+	if (out == NULL) {
+		return NULL;
+	}
+
+	if (_gdImageAvifCtx(im, out, quality, speed)) {
+		rv = NULL;
+	} else {
+		rv = gdDPExtractData(out, size);
+	}
+
+	out->gd_free(out);
+	return rv;
+}
+
+BGD_DECLARE(void *) gdImageAvifPtr (gdImagePtr im, int *size)
+{
+	return gdImageAvifPtrEx(im, size, DEFAULT_QUALITY, AVIF_SPEED_DEFAULT);
+}
+
+
+BGD_DECLARE(void) gdImageAvifCtx (gdImagePtr im, gdIOCtx * outfile, int quality, int speed)
+{
+  _gdImageAvifCtx(im, outfile, quality, speed);
+}
+
+/* 
+   We need this underscored function because gdImageAvifCtx() can't return anything.
+   And our functions that operate on a memory buffer need to know whether the encoding has succeeded.
+
+   If we're passed the DEFAULT_QUALITY of -1, set the quantizer params to DEFAULT_QUANTIZER
+
+   This function returns 0 on success, or 1 on failure.
+ */
+static avifBool _gdImageAvifCtx (gdImagePtr im, gdIOCtx * outfile, int quality, int speed)
+{
+  avifResult result;
+  avifRGBImage rgb;
+  avifRWData avifOutput = AVIF_DATA_EMPTY;
+  avifBool failed = AVIF_FALSE;
+
+  uint32_t val;
+  uint8_t * p;
+  uint8_t a;
+  int x, y;
+
+	if (im == NULL) {
+		return 1;
+	}
+
+	if (!gdImageTrueColor(im)) {
+		gd_error("avif doesn't support palette images");
+		return 1;
+	}
+
+  avifImage * avifIm = avifImageCreate(gdImageSX(im), gdImageSY(im), 8, DEFAULT_CHROMA_SUBSAMPLING);
+
+  rgb.width = gdImageSX(im);
+  rgb.height = gdImageSY(im);
+  rgb.depth = 8;
+  rgb.format = AVIF_RGB_FORMAT_RGBA;
+  rgb.chromaUpsampling = AVIF_CHROMA_UPSAMPLING_AUTOMATIC;
+  rgb.ignoreAlpha = AVIF_FALSE;
+  rgb.pixels = NULL;
+  avifRGBImageAllocatePixels(&rgb); // this allocates memory, and sets rgb.rowBytes and rgb.pixels
+
+  // Parse RGB data from the GD image, and copy it into the AVIF RGB image.
+  p = rgb.pixels;
+  for (y = 0; y < rgb.height; y++) {
+    for (x = 0; x < rgb.width; x++) {
+      val = im->tpixels[y][x];
+
+      a = convertTo8BitAlpha(gdTrueColorGetAlpha(val));
+
+      *(p++) = gdTrueColorGetRed(val);
+      *(p++) = gdTrueColorGetGreen(val);
+      *(p++) = gdTrueColorGetBlue(val);
+      *(p++) = a;
+    }
+  }
+
+  result = avifImageRGBToYUV(avifIm, &rgb);
+  if ((failed = isAvifError(result, "Could not convert image to YUV"))) {
```

Comment:
don't inline assignments into `if` statements

Reply:
Understood. It's so convenient. But not everyone likes it. (The compiler didn't until I added the extra layer of parens)